include/awkward/cuda-kernels/cuda_allocators.h

```diff
@@ -0,0 +1,128 @@
+  EXPORT_SYMBOL ERROR awkward_cuda_ptrU32_dealloc(const uint32_t* ptr);
+  EXPORT_SYMBOL ERROR awkward_cuda_ptr64_dealloc(const int64_t* ptr);
+  EXPORT_SYMBOL ERROR awkward_cuda_ptrU64_dealloc(const uint64_t* ptr);
+  EXPORT_SYMBOL ERROR awkward_cuda_ptrfloat32_dealloc(const float* ptr);
+  EXPORT_SYMBOL ERROR awkward_cuda_ptrfloat64_dealloc(const double* ptr);
+
+  EXPORT_SYMBOL ERROR awkward_cuda_H2Dbool(
+      bool** to_ptr,
+      bool* from_ptr,
+      int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_H2D8(
+      int8_t** to_ptr,
+      int8_t* from_ptr,
+      int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_H2DU8(
+        uint8_t** to_ptr,
+        uint8_t* from_ptr,
+        int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_H2D16(
+        int16_t** to_ptr,
+        int16_t* from_ptr,
+        int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_H2DU16(
+        uint16_t** to_ptr,
+        uint16_t* from_ptr,
+        int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_H2D32(
+        int32_t** to_ptr,
+        int32_t* from_ptr,
+        int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_H2DU32(
+        uint32_t** to_ptr,
+        uint32_t* from_ptr,
+        int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_H2D64(
+      int64_t** to_ptr,
+      int64_t* from_ptr,
+      int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_H2DU64(
+      uint64_t** to_ptr,
+      uint64_t* from_ptr,
+      int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_H2Dfloat32(
+      float** to_ptr,
+      float* from_ptr,
+      int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_H2Dfloat64(
+      double** to_ptr,
+      double* from_ptr,
+      int64_t length);
+
+  EXPORT_SYMBOL ERROR awkward_cuda_D2Hbool(
+    bool** to_ptr,
+    bool* from_ptr,
+    int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_D2H8(
+    int8_t** to_ptr,
+    int8_t* from_ptr,
+    int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_D2HU8(
+    uint8_t** to_ptr,
+    uint8_t* from_ptr,
+    int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_D2H16(
+    int16_t** to_ptr,
+    int16_t* from_ptr,
+    int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_D2HU16(
+    uint16_t** to_ptr,
+    uint16_t* from_ptr,
+    int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_D2H32(
+    int32_t** to_ptr,
+    int32_t* from_ptr,
+    int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_D2HU32(
+    uint32_t** to_ptr,
+    uint32_t* from_ptr,
+    int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_D2H64(
+    int64_t** to_ptr,
+    int64_t* from_ptr,
+    int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_D2HU64(
+    uint64_t** to_ptr,
+    uint64_t* from_ptr,
+    int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_D2Hfloat32(
+    float** to_ptr,
+    float* from_ptr,
+    int64_t length);
+  EXPORT_SYMBOL ERROR awkward_cuda_D2Hfloat64(
+    double** to_ptr,
+    double* from_ptr,
+    int64_t length);
```

Comment:
On the other hand, if these functions are hidden by the dispatch mechanism (as they should be, since the C++ codebase outside of the `kernel` namespace shouldn't be aware of CUDA vs main memory—that's what the `kernel::Lib` enum is for), then these internal functions can be named whatever you want and the externally visible ones, currently with names like `kernel::H2D<T>(...)`, could be `kernel::copy_to<L, T>(...)` or `kernel::copy_to<T>(kernel::Lib ptr_lib, ...)`, where the `kernel::Lib` parameter is either templated over (`L`) or runtime (`ptr_lib`). Then we don't need to know what "host" and "device" are, we're just copying it to something specified by the `kernel::Lib` enum.